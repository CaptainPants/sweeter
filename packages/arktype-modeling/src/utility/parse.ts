import { type ArkErrors, type } from 'arktype';
import { type AnyTypeConstraint } from '../type/index.js';

export type SafeParseResult<TArkType extends AnyTypeConstraint> =
    | { success: true; data: type.infer<TArkType> }
    | { success: false; summary: string; issues: ArkErrors };

export function safeParse<TArkType extends AnyTypeConstraint>(
    value: unknown,
    schema: TArkType,
): SafeParseResult<TArkType> {
    const res = schema(value);

    if (res instanceof type.errors) {
        return { success: false, summary: res.summary, issues: res.issues };
    }
    return { success: true, data: value as type.infer<TArkType> };
}

export function parse<TSchema extends AnyTypeConstraint>(
    value: unknown,
    schema: TSchema,
): type.infer<TSchema> {
    const res = schema(value);

    if (res instanceof type.errors) {
        throw new TypeError(res.summary);
    }
    return res as type.infer<TSchema>;
}

// TODO: not sure that there is an async model in ArkType (where Zod has one)
// so will have to see if these can have a meaningful implementation
export const parseAsync = (
    ...args: Parameters<typeof parse>
): Promise<ReturnType<typeof parse>> => {
    try {
        return Promise.resolve(parse(...args));
    } catch (err) {
        return Promise.reject(err);
    }
};

export const safeParseAsync = (
    ...args: Parameters<typeof safeParse>
): Promise<ReturnType<typeof safeParse>> => {
    try {
        return Promise.resolve(safeParse(...args));
    } catch (err) {
        return Promise.reject(err);
    }
};
