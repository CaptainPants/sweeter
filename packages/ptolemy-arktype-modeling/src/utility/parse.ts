import { type ArkErrors, type } from 'arktype';

import { type AnyTypeConstraint } from '../type/index.js';

export type SafeParseResult<TSchema extends AnyTypeConstraint> =
    | { success: true; data: type.infer<TSchema> }
    | { success: false; summary: string; issues: ArkErrors };

export function safeParse<TSchema extends AnyTypeConstraint>(
    value: unknown,
    schema: TSchema,
): SafeParseResult<TSchema> {
    const res = schema(value) as unknown;

    if (res instanceof type.errors) {
        return { success: false, summary: res.summary, issues: res.issues };
    }

    return { success: true, data: value as type.infer<TSchema> };
}

export function parse<TSchema extends AnyTypeConstraint>(
    value: unknown,
    schema: TSchema,
): type.infer<TSchema> {
    const res = schema(value) as unknown;

    if (res instanceof type.errors) {
        throw new TypeError(res.summary);
    }

    // eslint-disable-next-line @typescript-eslint/no-unsafe-return -- linter is choking
    return res as type.infer<TSchema>;
}

// TODO: not sure that there is an async model in ArkType (where Zod has one)
// so will have to see if these can have a meaningful implementation

export const parseAsync = async <TSchema extends AnyTypeConstraint>(
    value: unknown,
    schema: TSchema,
    // eslint-disable-next-line  @typescript-eslint/require-await -- We currently fake the async-ness of this, but will be investigating if there is a proper way to do it
): Promise<type.infer<TSchema>> => {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return -- linter is choking
    return parse(value, schema);
};

export const safeParseAsync = async <TSchema extends AnyTypeConstraint>(
    value: unknown,
    schema: TSchema,
    // eslint-disable-next-line  @typescript-eslint/require-await -- We currently fake the async-ness of this, but will be investigating if there is a proper way to do it
): Promise<SafeParseResult<TSchema>> => {
    return safeParse(value, schema);
};
