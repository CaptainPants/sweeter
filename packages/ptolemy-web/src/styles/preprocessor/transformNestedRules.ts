import {
    type AtRuleAstNode,
    type RuleAstNode,
    type RuleOrAtRule,
} from './types.js';

export function transformNestedRules(nodes: RuleOrAtRule[]): RuleOrAtRule[] {
    const result: RuleOrAtRule[] = [];

    transformMany(undefined, nodes, result, result);

    return result;
}

export function transformMany(
    parentSelectors: string[] | undefined,
    nodes: RuleOrAtRule[],
    currentRootOrAtRuleNestedNodes: RuleOrAtRule[],
    roots: RuleOrAtRule[],
): boolean {
    let anyProperties = false;

    for (const node of nodes) {
        const thisTime = transformOne(
            parentSelectors,
            node,
            currentRootOrAtRuleNestedNodes,
            roots,
        );
        anyProperties ||= thisTime;
    }

    return anyProperties;
}

export function transformOne(
    parentSelectors: string[] | undefined,
    node: RuleOrAtRule,
    currentRootOrAtRuleNestedNodes: RuleOrAtRule[],
    roots: RuleOrAtRule[],
): boolean {
    let anyProperties = false;

    if (node.$nodeType == 'at') {
        if (node.type == 'root') {
            // @root {  } basically resets & to nothing so you can write root rules within another rule for convenience
            if (node.nestedRules) {
                const thisTime = transformMany(
                    undefined,
                    node.nestedRules,
                    roots, // reset the currentRootOrAtRuleNestedNodes value for child nodes
                    roots,
                );
                anyProperties ||= thisTime;
            }

            // Properties don't make any sense here
        } else {
            // Create a new @ node
            const cloneChildren: RuleOrAtRule[] = [];
            const clone: AtRuleAstNode = {
                $nodeType: 'at',
                type: node.type,
                parameters: node.parameters,
                nestedRules: cloneChildren,
            };

            if (node.nestedRules) {
                const thisTime = transformMany(
                    parentSelectors,
                    node.nestedRules,
                    cloneChildren,
                    roots,
                );
                anyProperties ||= thisTime;
            }

            if (
                node.properties &&
                node.properties.length > 0 &&
                parentSelectors
            ) {
                cloneChildren.push({
                    $nodeType: 'rule',
                    selectors: parentSelectors,
                    properties: node.properties,
                    nestedRules: [],
                });
                anyProperties = true;
            }

            if (anyProperties) {
                currentRootOrAtRuleNestedNodes.push(clone);
            }
        }
    } else {
        const selectors = combineSelectors(parentSelectors, node.selectors);

        if (node.properties && node.properties.length > 0) {
            const newNode: RuleAstNode = {
                $nodeType: 'rule',
                selectors: selectors,
                properties: node.properties,
                nestedRules: [],
            };

            currentRootOrAtRuleNestedNodes.push(newNode);

            anyProperties = true;
        }

        if (node.nestedRules && node.nestedRules.length > 0) {
            const thisTime = transformMany(
                selectors,
                node.nestedRules,
                currentRootOrAtRuleNestedNodes,
                roots,
            );
            anyProperties ||= thisTime;
        }
    }

    return anyProperties;
}

function combineSelectors(
    parentSelectors: string[] | undefined,
    selectors: string[],
): string[] {
    if (parentSelectors) {
        const result: string[] = [];

        // if :is() is supported, we don't need to have an N^2 here, which leads
        // to N^M complexity (with M being the depth of nest). We can just do:
        // is(<parentSelector>) childSelector. This will break using & to make a
        // new selector though.
        for (const outerSelector of parentSelectors) {
            for (const innerSelector of selectors) {
                const selector = outerSelector
                    ? innerSelector.includes('&')
                        ? innerSelector.replaceAll('&', outerSelector)
                        : outerSelector + ' ' + innerSelector
                    : innerSelector;

                result.push(selector);
            }
        }

        return result;
    } else {
        return selectors;
    }
}
