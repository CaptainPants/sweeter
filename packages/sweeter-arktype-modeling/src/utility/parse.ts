import { type ArkErrors, type } from 'arktype';

import { type AnyTypeConstraint } from '../type/index.js';

export type SafeParseResult<TSchema extends AnyTypeConstraint> =
    | { success: true; data: type.infer<TSchema> }
    | { success: false; summary: string; issues: ArkErrors };

export function safeParse<TSchema extends AnyTypeConstraint>(
    value: unknown,
    schema: TSchema,
): SafeParseResult<TSchema> {
    const res = schema(value);

    if (res instanceof type.errors) {
        return { success: false, summary: res.summary, issues: res.issues };
    }
    return { success: true, data: value as type.infer<TSchema> };
}

export function parse<TSchema extends AnyTypeConstraint>(
    value: unknown,
    schema: TSchema,
): type.infer<TSchema> {
    const res = schema(value);

    if (res instanceof type.errors) {
        throw new TypeError(res.summary);
    }
    return res;
}

// TODO: not sure that there is an async model in ArkType (where Zod has one)
// so will have to see if these can have a meaningful implementation
 
export const parseAsync = async <TSchema extends AnyTypeConstraint>(
    value: unknown,
    schema: TSchema,
    // eslint-disable-next-line  @typescript-eslint/require-await -- We currently fake the async-ness of this, but will be investigating if there is a proper way to do it
): Promise<type.infer<TSchema>> => {
    return parse(value, schema);
};

export const safeParseAsync = async <TSchema extends AnyTypeConstraint>(
    value: unknown,
    schema: TSchema,
    // eslint-disable-next-line  @typescript-eslint/require-await -- We currently fake the async-ness of this, but will be investigating if there is a proper way to do it
): Promise<SafeParseResult<TSchema>> => {
    return safeParse(value, schema);
};
