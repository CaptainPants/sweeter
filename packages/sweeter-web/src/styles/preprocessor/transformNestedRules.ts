import { assertNeverNullish } from '@captainpants/sweeter-core';
import {
    type RuleAstNode,
    type AtRuleAstNode,
    type RuleOrAtRule,
    type PropertyAstNode,
} from './types.js';

export function transformNestedRules(ast: RuleOrAtRule[]): RuleOrAtRule[] {
    const result: RuleOrAtRule[] = [];

    for (const item of ast) {
        transform([], [], item, result);
    }

    return result;
}

export function transform(
    atRulePath: AtRuleAstNode[],
    rulePath: RuleAstNode[],
    item: RuleOrAtRule,
    roots: RuleOrAtRule[],
): void {
    // TODO: optimise for @ rule containing regular rules - currently there is a clone of the @ rule for each

    if (item.$nodeType == 'at' && item.type == 'root') { // @root {  } basically rests & to nothing so you can write root rules within another rule for convenience
        if (item.nestedRules) {
            for (const inner of item.nestedRules) {
                transform([], [], inner, roots);
            }
        }
        return;
    }

    const newAtRulePath =
        item.$nodeType == 'at' ? [...atRulePath, item] : atRulePath;
    const newRulePath =
        item.$nodeType == 'rule' ? [...rulePath, item] : rulePath;

    // Add a node to contain the properties in this (possibly nested) node
    if (item.properties && item.properties.length > 0) {
        roots.push(createRuleFor(newAtRulePath, newRulePath, item.properties));
    }

    if (item.nestedRules) {
        for (const nestedRule of item.nestedRules) {
            transform(newAtRulePath, newRulePath, nestedRule, roots);
        }
    }
}

function createRuleFor(
    atRulePath: AtRuleAstNode[],
    rulePath: RuleAstNode[],
    properties: PropertyAstNode[] | undefined,
): RuleOrAtRule {
    let result: RuleOrAtRule = {
        $nodeType: 'rule',
        selectors: [expandSelectorsForPath(rulePath)],
        nestedRules: [],
        properties: properties ?? [],
    };

    // Wrap all @ rules around the rule
    for (let i = atRulePath.length - 1; i >= 0; --i) {
        const source = atRulePath[i]!;

        result = {
            $nodeType: 'at',
            type: source.type,
            parameters: source.parameters,
            nestedRules: [result],
        };
    }

    return result;
}

function expandSelectorsForPath(nodePath: RuleAstNode[]): string {
    const results: string[] = [];
    expandSelectorsForPathImplementation(undefined, nodePath, 0, results);
    return results.join(', ');
}

function expandSelectorsForPathImplementation(
    compiledParentSelector: string | undefined,
    nodePath: RuleAstNode[],
    nodePathIndex: number,
    output: string[],
): void {
    if (nodePathIndex >= nodePath.length) {
        assertNeverNullish(compiledParentSelector);
        output.push(compiledParentSelector);
        return;
    }

    const firstPart = nodePath[nodePathIndex]!;

    for (const current of firstPart.selectors) {
        const selector = compiledParentSelector
            ? current.includes('&')
                ? current.replace('&', compiledParentSelector)
                : compiledParentSelector + ' ' + current
            : current;

        expandSelectorsForPathImplementation(
            selector,
            nodePath,
            nodePathIndex + 1,
            output,
        );
    }
}
