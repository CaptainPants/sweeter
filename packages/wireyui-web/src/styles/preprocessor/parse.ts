import { indexOfAny } from './internal/indexOfAny.js';
import type {
    AtRuleAstNode,
    PropertyAstNode,
    RuleAstNode,
    RuleBodyParts,
    RuleOrAtRule,
} from './types.js';

/**
 * Parse CSS content of a file
 * @param css
 * @returns
 */
export function parse(css: string): RuleOrAtRule[] {
    const parser = new Parser(css);
    return parser.parse();
}

/**
 * Parse CSS content of a class.
 * @param css
 * @returns
 */
export function parseClassContent(css: string): RuleBodyParts {
    const parser = new Parser(css);
    return parser.parseRuleBodyContent();
}

const brace = ['{'.charCodeAt(0)] as const;
const semiOrBrace = ['{'.charCodeAt(0), ';'.charCodeAt(0)] as const;
const semiOrBraceOrCloseBrace = [
    '{'.charCodeAt(0),
    ';'.charCodeAt(0),
    '}'.charCodeAt(0),
] as const;

class Parser {
    constructor(input: string) {
        this.#input = input;
        this.#index = 0;
    }

    #input: string;
    #index: number;

    public parse(): RuleOrAtRule[] {
        return this.#parseRuleOrAtRule(false);
    }

    #parseRuleOrAtRule(inBlock: boolean): RuleOrAtRule[] {
        const result: RuleOrAtRule[] = [];

        this.#skipWhiteSpace();

        while (this.#index < this.#input.length) {
            // whitespace should already have been skipped

            const next = this.#input[this.#index];
            if (next === '}') {
                if (inBlock) {
                    // move #index past the block
                    this.#index += 1;
                    break;
                } else {
                    throw new Error(this.#errorMessage('Unexpected'));
                }
            } else if (next === '@') {
                result.push(this.#parseAtRule());
            } else {
                result.push(this.#parseRule());
            }

            this.#skipWhiteSpace();
        }

        return result;
    }

    #parseRule(): RuleAstNode {
        const startOfBody = indexOfAny(brace, this.#input, this.#index);
        if (!startOfBody)
            throw new Error(this.#errorMessage('No block body found'));

        const selector = this.#input.substring(this.#index, startOfBody);

        this.#index = startOfBody + 1;
        const { nestedRules, properties } = this.parseRuleBodyContent();
        this.#index += 1;

        return {
            $type: 'rule',
            selector,
            nestedRules,
            properties,
        };
    }

    /**
     * Positions the #index at the next index on completion
     * @returns
     */
    #parseAtRule(): AtRuleAstNode {
        ++this.#index; // move past @

        this.#skipWhiteSpace();

        // The opening line must end with a ; or a {
        const foundIndex = indexOfAny(semiOrBrace, this.#input, this.#index);
        if (foundIndex === undefined)
            throw new Error(
                this.#errorMessage('Could not find end of @ rule preamble as.'),
            );

        const text = this.#input.substring(this.#index, foundIndex);

        const whatDidWeFind = this.#input[foundIndex]!;

        // position after the ; or {
        this.#index = foundIndex + 1;

        if (whatDidWeFind === ';') {
            return { $type: 'at', text: text };
        } else {
            // position at start of block
            this.#index = foundIndex;
            const block = this.#parseRuleOrAtRule(true);

            return {
                $type: 'at',
                text: text,
                body: block,
            };
        }
    }

    /**
     * Expects #index to be pointing to the firat character in the block (after the {)
     * Positions the #index at the closing brace (or end of file) on completion
     */
    parseRuleBodyContent(): RuleBodyParts {
        const nestedRules: RuleAstNode[] = [];
        const properties: PropertyAstNode[] = [];

        while (this.#index < this.#input.length) {
            this.#skipWhiteSpace();

            if (this.#input[this.#index] === '}') {
                break;
            }

            const foundSemiOrBlockStartOrBlockEndIndex = indexOfAny(
                semiOrBraceOrCloseBrace,
                this.#input,
                this.#index,
            );

            if (foundSemiOrBlockStartOrBlockEndIndex === undefined) {
                // I think this is really if the file ends during a rule
                const wholeProperty = this.#input.substring(this.#index);

                if (wholeProperty.trim().length > 0) {
                    properties.push({
                        $type: 'property',
                        text: wholeProperty,
                    });
                    this.#index = this.#input.length; // end of file
                }
                break;
            }

            const which = this.#input[foundSemiOrBlockStartOrBlockEndIndex];

            // nested rule
            if (which === '{') {
                const nestedRule = this.#parseRule(); // positions #index at the next character
                nestedRules.push(nestedRule);
            } else {
                const wholeProperty = this.#input.substring(
                    this.#index,
                    foundSemiOrBlockStartOrBlockEndIndex,
                );
                properties.push({
                    $type: 'property',
                    text: wholeProperty,
                });
                this.#index = foundSemiOrBlockStartOrBlockEndIndex + 1;
            }
        }

        return { nestedRules, properties };
    }

    /**
     * Position #index at the next non-whitespace character
     */
    #skipWhiteSpace(): void {
        while (
            this.#index < this.#input.length &&
            ' \t\n\r\v'.indexOf(this.#input[this.#index]!) >= 0
        ) {
            ++this.#index;
        }
    }

    /**
     * Create an error object with useful positional information from #index
     * @param message
     * @returns
     */
    #errorMessage(message: string): string {
        const upToIndex = this.#input.substring(0, this.#index);

        const lineNumber = upToIndex.match(/\r?\n/g)?.length ?? 0;

        let columnIndex = 0;

        for (
            let countBackwardsIndex = Math.min(
                this.#index,
                this.#input.length - 1,
            );
            countBackwardsIndex >= 0;
            --countBackwardsIndex
        ) {
            if (this.#input[countBackwardsIndex]!.match(/^\r?\n/g)) {
                break;
            }

            ++columnIndex;
        }

        return `At index ${this.#index} (line: ${lineNumber + 1}, column: ${
            columnIndex + 1
        }) [${this.getIndexContextString()}]: ${message}`;
    }

    toString() {
        return `${this.#index} -> ${this.getIndexContextString()}`;
    }

    getIndexContextString() {
        const numberOfCharsOnEitherSide = 20;

        const from = Math.max(this.#index - numberOfCharsOnEitherSide, 0);

        if (this.#index >= this.#input.length) {
            // this #index is already off the end, so there is no 'current'
            return `${this.#input.substring(from, this.#index)} EOF`;
        }

        let to = this.#index + numberOfCharsOnEitherSide + 1;
        let atEnd: boolean = false;

        if (to >= this.#input.length) {
            to = this.#input.length - 1;
            atEnd = true;
        }

        const result = `${this.#input.substring(from, this.#index)} <|${
            this.#input[this.#index]
        }|> ${this.#input.substring(this.#index + 1, to)}${
            atEnd ? ' EOF' : ''
        }`;

        return result;
    }
}
