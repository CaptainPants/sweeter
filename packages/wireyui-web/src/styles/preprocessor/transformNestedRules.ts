import { splitByUnparenthesizedCommas } from './internal/splitByUnparenthesizedCommas.js';
import { type AtRuleAstNode, type RuleOrAtRule } from './types.js';

export function transformNestedRules(ast: RuleOrAtRule[]): RuleOrAtRule[] {
    const result: RuleOrAtRule[] = [];

    for (const item of ast) {
        transform(undefined, item, result);
    }

    return result;
}

export function transform(
    selector: string | undefined,
    item: RuleOrAtRule,
    roots: RuleOrAtRule[],
): void {
    // Note that this should only happen at the root
    if (item.$type === 'at') {
        const newNode: AtRuleAstNode = {
            $type: 'at',
            text: item.text,
        };

        if (item.body && item.body.length > 0) {
            const body: RuleOrAtRule[] = [];
            for (const childItem of item.body) {
                transform(undefined, childItem, body);
            }
            item.body = body;
        }

        roots.push(newNode);
        return;
    }

    const selectorParts = splitByUnparenthesizedCommas(item.selector);

    // Add a node to contain the properties in this (possibly nested) node
    if (item.properties.length > 0) {
        if (!selector) {
            throw new Error('This is probably a bug');
        }

        for (const part of selectorParts) {
            roots.push({
                $type: 'rule',
                selector: combineSelector(selector, part),
                nestedRules: [],
                properties: item.properties,
            });
        }
    }

    if (item.nestedRules) {
        for (const part of selectorParts) {
            for (const nestedRule of item.nestedRules) {
                // E.g. .class1, .class2, :not(.class3) will have 3 elements
                transform(combineSelector(selector, part), nestedRule, roots);
            }
        }
    }
}

function combineSelector(
    parentSelector: string | undefined,
    selector: string,
): string {
    if (parentSelector) return `${parentSelector} ${selector}`;
    return selector;
}
