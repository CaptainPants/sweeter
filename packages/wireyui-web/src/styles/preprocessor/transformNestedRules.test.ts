import { parse } from './parse.js';
import { transformNestedRules } from './transformNestedRules.js';

it('test1', () => {
    const input = `
        .outer1 {
            .inner1 {
                color: red;
            }
        }
    `;

    const parsed = parse(input);

    expect(parsed).toMatchSnapshot();

    const transformed = transformNestedRules(parsed);

    expect(transformed).toMatchSnapshot();
});

it('test2', () => {
    const input = `
        .outer1, .outer2 {
            .inner1, .inner2 {
                color: red;
            }
        }
    `;

    const parsed = parse(input);

    expect(parsed).toMatchSnapshot();

    const transformed = transformNestedRules(parsed);

    expect(transformed).toMatchSnapshot();
});

it('test3', () => {
    const input = `
        .outer1, .outer2 {
            color: blue;

            .inner1, .inner2 {
                color: red;
            }
        }
    `;

    const parsed = parse(input);

    expect(parsed).toMatchSnapshot();

    const transformed = transformNestedRules(parsed);

    expect(transformed).toMatchSnapshot();
});

it('complex selectors', () => {
    const input = `
        .outer1, .outer2 {
            color: blue;

            .inner1, .inner2 {
                color: red;
            }
        }
    `;

    const parsed = parse(input);

    expect(parsed).toMatchSnapshot();

    const transformed = transformNestedRules(parsed);

    expect(transformed).toMatchSnapshot();
});
