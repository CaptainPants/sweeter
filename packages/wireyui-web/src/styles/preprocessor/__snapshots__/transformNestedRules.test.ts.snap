// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test2 1`] = `
[
  {
    "$type": "rule",
    "nestedRules": [],
    "properties": [
      {
        "$type": "property",
        "text": "color: red",
      },
    ],
    "selector": ".outer1 .inner1",
  },
  {
    "$type": "rule",
    "nestedRules": [],
    "properties": [
      {
        "$type": "property",
        "text": "color: red",
      },
    ],
    "selector": ".outer1  .inner2 ",
  },
  {
    "$type": "rule",
    "nestedRules": [],
    "properties": [
      {
        "$type": "property",
        "text": "color: red",
      },
    ],
    "selector": " .outer2  .inner1",
  },
  {
    "$type": "rule",
    "nestedRules": [],
    "properties": [
      {
        "$type": "property",
        "text": "color: red",
      },
    ],
    "selector": " .outer2   .inner2 ",
  },
]
`;
